weave/flow/dataset.py

```diff
@@ -181,3 +199,178 @@ def select(self, indices: Iterable[int]) -> Self:
+            ds_with_length = ds.map(add_text_length)
+            ds_combined = ds.map(combine_fields)
+            ds_categorized = ds.map(classify_language)
+
+            # Example output for ds_with_length[0]:
+            # {
+            #     "id": 1,
+            #     "text": "Hello world",
+            #     "lang": "en",
+            #     "length": 11
+            # }
+            ```
+        """
+        processed_rows = []
+
+        # Inspect the function signature
+        sig = inspect.signature(func)
+        param_names = list(sig.parameters.keys())
+
+        # Check if a lambda function is being used and warn about traceability
+        if func.__name__ == "<lambda>":
+            warnings.warn(
+                "Lambda functions in Dataset.map reduce traceability in Weave. "
+                "For better tracking and visualization in the UI, use named functions instead. "
+                "Example: `def process(x): return {'result': x*2}` instead of `lambda x: {'result': x*2}`",
+                UserWarning,
+                stacklevel=2,
+            )
+
+        async def process_row(row: dict) -> dict:
+            try:
+                # Extract and pass specific parameters
+                kwargs = {}
+                for name in param_names:
+                    if name in row:
+                        kwargs[name] = row[name]
+                    else:
+                        # Skip this parameter or use a default if available
+                        if sig.parameters[name].default is not inspect.Parameter.empty:
+                            kwargs[name] = sig.parameters[name].default
+                        else:
+                            raise ValueError(
+                                f"Function expects parameter '{name}' but this column "
+                                f"is not in the dataset row: {short_str(row)}"
+                            )
+
+                result = func(**kwargs)
+
+                # If the user function is async, await it
+                if asyncio.iscoroutine(result):
+                    result = await result
+
+                # Handle the result based on its type
+                if not isinstance(result, dict):
+                    # Raise error for non-dictionary returns
+                    raise TypeError(
+                        f"Function must return a dictionary, but got {type(result).__name__} "
+                        f"from function '{func.__name__}'. Use a dictionary to specify "
+                        f"which columns to add or update."
+                    )
+            except Exception as e:
+                # Log the error and propagate it to stop the map operation
+                print(f"Error processing row with map function: {short_str(row)}")
+                traceback.print_exc()
+                raise e
+            else:
+                # Update the original row with the results
+                row_copy = row.copy()
```

Comment:
This is a shallow copy -- is that intentional?

Reply:
nope!